pkg/services/provisioning/dashboards/file_reader.go

```diff
@@ -412,42 +413,139 @@ type dashboardIdentity struct {
+	uidUsage   map[string]uint8
+	titleUsage map[dashboardIdentity]uint8
 }
 
-func (checker provisioningSanityChecker) track(pm provisioningMetadata) {
+func (t *usageTracker) track(pm provisioningMetadata) {
 	if len(pm.uid) > 0 {
-		checker.uidUsage[pm.uid]++
+		t.uidUsage[pm.uid]++
 	}
 	if pm.identity.Exists() {
-		checker.titleUsage[pm.identity]++
+		t.titleUsage[pm.identity]++
 	}
 }
 
-func (checker provisioningSanityChecker) logWarnings(log log.Logger) {
-	for uid, times := range checker.uidUsage {
-		if times > 1 {
-			log.Error("the same 'uid' is used more than once", "uid", uid, "provider", checker.provisioningProvider)
+type set struct {
+	data map[string]struct{}
+}
+
+func newSet() *set {
+	return &set{make(map[string]struct{})}
+}
+
+func (s *set) Add(entry string) {
+	s.data[entry] = struct{}{}
+}
+
+func (s *set) AsSlice() []string {
+	entries := make([]string, len(s.data))
+	index := 0
+
+	for entry := range s.data {
+		entries[index] = entry
+		index++
+	}
+
+	return entries
+}
+
+type duplicate struct {
+	Readers *set
+	Sum     uint8
+}
+
+func newDuplicate() *duplicate {
+	return &duplicate{Readers: newSet()}
+}
+
+type duplicateEntries struct {
+	Title map[dashboardIdentity]*duplicate
+	UID   map[string]*duplicate
+}
+
+func newDuplicateEntries() *duplicateEntries {
+	return &duplicateEntries{Title: make(map[dashboardIdentity]*duplicate), UID: make(map[string]*duplicate)}
+}
+
+type sanityChecker struct {
+	mux      sync.RWMutex
+	trackers map[string]*usageTracker
+}
+
+func newSanityChecker() *sanityChecker {
+	return &sanityChecker{trackers: make(map[string]*usageTracker)}
+}
+
+func (c *sanityChecker) Set(readerName string, tracker *usageTracker) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	c.trackers[readerName] = tracker
+}
+
+func (c *sanityChecker) getDuplicates() *duplicateEntries {
+	duplicates := newDuplicateEntries()
+
+	for readerName, tracker := range c.trackers {
+		for uid, times := range tracker.uidUsage {
+			if _, ok := duplicates.UID[uid]; !ok {
+				duplicates.UID[uid] = newDuplicate()
+			}
+			duplicates.UID[uid].Sum += times
+			duplicates.UID[uid].Readers.Add(readerName)
+		}
+
+		for id, times := range tracker.titleUsage {
+			if _, ok := duplicates.Title[id]; !ok {
+				duplicates.Title[id] = newDuplicate()
+			}
+			duplicates.Title[id].Sum += times
+			duplicates.Title[id].Readers.Add(readerName)
+		}
+	}
+
+	return duplicates
+}
+
+func (c *sanityChecker) logWarnings(log log.Logger) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	duplicates := c.getDuplicates()
+
+	for uid, usage := range duplicates.UID {
+		if usage.Sum > 1 {
+			log.Error("the same 'uid' is used more than once", "uid", uid, "times", usage.Sum, "providers", usage.Readers.AsSlice())
+		}
+	}
+
+	for id, usage := range duplicates.Title {
+		if usage.Sum > 1 {
+			log.Error("the same 'folder/title' is used more than once", "title", id.title, "folderID", id.folderID, "times", usage.Sum, "providers", usage.Readers.AsSlice())
```

Comment:
What is a folder/title? Is this supposed to be a folder title? I honestly don't know, which is why I have to ask.
```suggestion
			log.Error("the same 'folder/title' is used more than once", "title", id.title, "folderID", id.folderID, "times", usage.Sum, "providers", usage.Readers.AsSlice())
```

Reply:
If I understand this correctly, maybe something like this would be clearer?

```suggestion
			log.Error("dashboard title is not unique in folder", "title", id.title, "folderID", id.folderID, "times", usage.Sum, "providers", usage.Readers.AsSlice())
```

Reply:
Same as @sakjur, if it's actually about dashboard title's uniqueness within folder, this is a much better error message.

Reply:
I will use the one you provided. Thanks.